RTL direction support via properties panel (revised)

okaeiz opened the pull request:
Has worked on https://github.com/bpmn-io/form-js/issues/1205

 This PR (within the the first commit) adds a new form-js entry named `DirectionEntry` to the properties panel `AppearanceGroup` that is activated for the form component (default type).


![6](https://github.com/bpmn-io/form-js/assets/100111649/f1f0b7da-e481-457a-9871-256962951469)


It is also intended to modify form components to adhere to the direction setting that is selected from the `DirectionEntry`.

renovate[bot] committed:
chore: update dependency @rollup/plugin-commonjs to v26

okaeiz committed:
Implemented RTL adherence for Action and Selection components

okaeiz committed:
Added DirectionEntry to the AppearanceGroup

okaeiz commented:
Here's the demo for Action and Selection components:

https://github.com/bpmn-io/form-js/assets/100111649/6a4fe182-546e-4e7b-925c-4a83f84ce331

okaeiz committed:
Implemented RTL adherence for Input and Presentation and Containers components

okaeiz commented:
Here's the demo for Input, Presentation and Containers components:

https://github.com/bpmn-io/form-js/assets/100111649/d6f69f39-83c0-40d5-828e-5ec3ab8e0bf6

okaeiz commented:
This looks completed to me. Here's a recap:

- I created a new PR with 3 commits and less diff:
  - Properties panel and DirectionEntry
  - Input, Presentation, and Containers components
  - Action and Selection components

I did **NOT** change the HTML View because its direction can already be changed via CSS attributes.  
I did **NOT** change iframe, separator, and spacer.  
I did **NOT** change expression because it's always LTR.
I did **NOT** include any merge commits as you requested.

@Skaiir I'd appreciate it if you take a look at this.

Skaiir commented:
Currently working on a review of this @okaeiz. Amazing work already, although there will need to be a few things changed before we can approve :D. But I'm confident together we'll get this into the library rather soon ;)

Skaiir commented on packages/form-js-viewer/src/render/components/form-fields/Button.js:
```diff
@@ -1,17 +1,25 @@
+      class={formFieldClasses(type)}
+      style={{
+        direction: direction,
+        fontFamily: 'Vazirmatn, sans-serif',
```
We have a css variable that defines fontFamily, so you can set it across the entire application rather easily (in general you'll find we don't use inline styles much, except for alignment). So **if** we're implementing it within the form-js, we should do it in one place only (somewhere in the form root).

For now, because I still need to understand a little bit more about the font stuff, please just remove it. We'll get back to it when everything else is done.

PS: the comments on this file apply across the board.

okaeiz replied:
✅

okaeiz replied:
I removed the font; but I kindly ask you to reconsider this decision. The font makes no difference to English and the languages that have a similar alphabet to English. But in Persian, it is one of the main (if not the main) reasons why people decide not to use a software.

If you take a look at Github, you'll find out that Github is also using [Vazirmatn](https://github.com/rastikerdar/vazirmatn) for its Persian strings. It is highly admired among open-source community.

Skaiir reviewed (changes requested):
I've mentioned a few things, but overall the direction is very good. There are a few things missing, although we don't need to tackle everything (but would be nice):

- [ ] Unimplemented
	- [ ] HTML should also be covered as a default (if it is relatively simple), yes you can define it within the code, but having a single configuration handle everything is just generally good.
	- [ ] The field placeholders (1*) should probably be RTL as well, just for visual consistency. But this isn't super important.
	- [ ] Iframe title should also be aligned (iframe content shouldn't be modified though, that is correct)
- [ ] Issues
	- [ ] Delete icon overlapping label (2*), would flip the button alignment
	- [ ] Separator for the number field on the wrong side (2*) (should be flipped)
	- [ ] Spacing between button and text in dynamic lists are squashed (4*), they might need some cleaning up

1* 
![grafik](https://github.com/bpmn-io/form-js/assets/17801113/f919891b-1c33-4b55-8a13-6f564e04666e)

2*
![grafik](https://github.com/bpmn-io/form-js/assets/17801113/6557c1b3-bdc6-4b5b-a3d4-1ea9c75deeca)

3*
![grafik](https://github.com/bpmn-io/form-js/assets/17801113/3f252710-a34a-42b0-9f6c-d279bd19c464)

4*
![grafik](https://github.com/bpmn-io/form-js/assets/17801113/b4cb1d42-f05c-46d0-9abc-0140639903c2)

For these smaller things, what will be needed is to make adjustments to the styles. So what I would do is, based on the direction setting, set a new class to the root of the form renderer. Then we'd modify the CSS using this tag to adjust those small things.

As mentioned, regarding the fonts, we can discuss this at the end.

Also, ensure that you pass our CI via the npm run all command. It'll show you any issues with the code. 

It's a lot of text on my part, but it shouldn't be a ton of changes. You don't need to close the PR, just force push your fixes to the PR.

Nice work so far, this will be a very useful change.

okaeiz commented on packages/form-js-editor/src/features/properties-panel/entries/DirectionEntry.js:
```diff
@@ -0,0 +1,61 @@
+
+function Direction(props) {
+  const { editField, field, id } = props;
+  const { setDirection } = useDirection(); // Get the context
```
✅

okaeiz commented on packages/form-js-editor/src/features/properties-panel/entries/DirectionEntry.js:
```diff
@@ -0,0 +1,61 @@
+  };
+
+  const setValue = (value) => {
+    setDirection(value); // Update context
```
✅

okaeiz commented on packages/form-js-editor/src/features/properties-panel/entries/index.js:
```diff
@@ -39,3 +39,5 @@ export { RowCountEntry } from './RowCountEntry';
 export { HeadersSourceSelectEntry } from './HeadersSourceSelectEntry';
 export { ColumnsExpressionEntry } from './ColumnsExpressionEntry';
 export { StaticColumnsSourceEntry } from './StaticColumnsSourceEntry';
+export { DirectionEntry } from './DirectionEntry';
```
✅

okaeiz commented on packages/form-js-editor/src/features/properties-panel/groups/AppearanceGroup.js:
```diff
@@ -1,10 +1,13 @@
     ...AdornerEntry({ field, editField }),
     ...GroupAppearanceEntry({ field, editField }),
     ...LayouterAppearanceEntry({ field, editField }),
+    ...LayouterAppearanceEntry({ field, editField }),
```
✅

okaeiz commented on packages/form-js-editor/src/features/properties-panel/entries/DirectionEntry.js:
```diff
@@ -0,0 +1,61 @@
+import { get } from 'min-dash';
+import { useService } from '../hooks';
+import { SelectEntry, isSelectEntryEdited } from '@bpmn-io/properties-panel';
+import { useDirection } from '../../../../../form-js-viewer/src/render/context/DirectionContext';
```
✅

okaeiz committed:
fixed some of the reviewer issues

okaeiz commented:
I made a new commit. Here's a recap:

- Implemented
  - I covered HTML View to adhere to direction.
  - I covered iFrame to adhere to direction.
 
- Issues fixed
  - Icon overlapping for DynamicList is fixed both in Editor and Viewer
  - Separator for the number field is now on the right side
  - Removed unnecessary comments
  - Removed the font attribute (temporarily) ---> will be discussed


I did **NOT** change the Expression field because whenever we have an LTR field in the RTL direction (such as expressions, emails, URLs, etc.) the placeholder should also be LTR; otherwise it will be confusing for the users. 
The code is now formatted with Prettier and eslint. I tried my best to adhere to the formatting guidelines.

@Skaiir I'd appreciate it if you take a look at this.

okaeiz commented:
Any updates?

Skaiir commented:
Hey @okaeiz, sorry we haven't been able to be around, unfortunately things got pretty busy pretty fast. I do have some amount of time now so I'll catch up with this today.

Skaiir commented:
Apologies for the delays it looks like this won't get taken a look at for another week at least. Unfortunately some other priorities came up.

okaeiz commented:
Take your time. I'll be available for further modifications.

okaeiz commented:
What is the status of this PR? Should I give up already?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/bpmn-io/form-js?pullRequest=1213) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/bpmn-io/form-js?pullRequest=1213) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/bpmn-io/form-js?pullRequest=1213) it.</sub>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/bpmn-io/form-js?pullRequest=1213) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/bpmn-io/form-js?pullRequest=1213) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/bpmn-io/form-js?pullRequest=1213) it.</sub>
